driver-core/src/main/com/mongodb/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -0,0 +1,160 @@
+import static com.mongodb.operation.OperationHelper.LOGGER;
+
+final class AsyncChangeStreamBatchCursor<T> implements AsyncBatchCursor<T> {
+    private final AsyncReadBinding binding;
+    private final ChangeStreamOperation<T> changeStreamOperation;
+
+    private volatile BsonDocument resumeToken;
+    private volatile AsyncBatchCursor<RawBsonDocument> wrapped;
+
+    AsyncChangeStreamBatchCursor(final ChangeStreamOperation<T> changeStreamOperation,
+                                 final AsyncBatchCursor<RawBsonDocument> wrapped,
+                                 final AsyncReadBinding binding) {
+        this.changeStreamOperation = changeStreamOperation;
+        this.resumeToken = changeStreamOperation.getResumeToken();
+        this.wrapped = wrapped;
+        this.binding = binding;
+        binding.retain();
+    }
+
+    AsyncBatchCursor<RawBsonDocument> getWrapped() {
+        return wrapped;
+    }
+
+    @Override
+    public void next(final SingleResultCallback<List<T>> callback) {
+        resumeableOperation(new AsyncBlock() {
+            @Override
+            public void apply(final AsyncBatchCursor<RawBsonDocument> cursor, final SingleResultCallback<List<RawBsonDocument>> callback) {
+                cursor.next(callback);
+            }
+        }, convertResultsCallback(callback));
+    }
+
+    @Override
+    public void tryNext(final SingleResultCallback<List<T>> callback) {
+        resumeableOperation(new AsyncBlock() {
+            @Override
+            public void apply(final AsyncBatchCursor<RawBsonDocument> cursor, final SingleResultCallback<List<RawBsonDocument>> callback) {
+                cursor.tryNext(callback);
+            }
+        }, convertResultsCallback(callback));
+    }
+
+    @Override
+    public void setBatchSize(final int batchSize) {
+        wrapped.setBatchSize(batchSize);
+    }
+
+    @Override
+    public int getBatchSize() {
+        return wrapped.getBatchSize();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return wrapped.isClosed();
+    }
+
+    @Override
+    public void close() {
+        wrapped.close();
+        binding.release();
+    }
+
+    private interface AsyncBlock {
+        void apply(AsyncBatchCursor<RawBsonDocument> cursor, SingleResultCallback<List<RawBsonDocument>> callback);
+    }
+
+    private void resumeableOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback) {
+        asyncBlock.apply(wrapped, new SingleResultCallback<List<RawBsonDocument>>() {
+            @Override
+            public void onResult(final List<RawBsonDocument> result, final Throwable t) {
+                if (t == null) {
+                    callback.onResult(result, null);
+                } else if (t instanceof MongoNotPrimaryException
+                        || t instanceof MongoCursorNotFoundException
+                        || t instanceof MongoSocketReadException) {
```

Comment:
This should be MongoSocketException. Otherwise we exclude socket open and socket write exceptions.